applications/crossbar/src/modules/provisioner_v5.erl

```diff
@@ -113,21 +98,74 @@ update_account(AccountId, JObj, AuthToken) ->
-%% @private
-%% @doc
-%%
-%% @end
-%%--------------------------------------------------------------------
--spec provision_data(wh_json:object()) -> wh_json:object().
-provision_data(JObj) ->
-    Routines =
-        [fun set_realm/1
-         ,fun set_owner/1
-         ,fun maybe_set_timezone/1
-         ,fun create_provision_settings/1
-        ],
-    lists:foldl(fun(F, J) -> F(J) end, JObj, Routines).
+%% @public
+-spec update_user(ne_binary(), wh_json:object(), ne_binary()) -> 'ok'.
+update_user(AccountId, JObj, AuthToken) ->
+    case wh_json:get_value(<<"pvt_type">>, JObj) of
+        <<"user">> ->
+            save_user(AccountId, JObj, AuthToken);
+        <<"device">> ->
+            save_user(AccountId, JObj, AuthToken);
+        _ -> ok %% Gets rid of VMbox
+    end.
+
+-spec save_user(ne_binary(), wh_json:object(), ne_binary()) -> 'ok'.
+save_user(AccountId, JObj, AuthToken) ->
+    Realm   = wh_json:get_value(<<"realm">>, set_realm(JObj)),
+    TZ      = wh_json:get_value(<<"timezone">>, JObj),
+    OwnerId = wh_json:get_value(<<"id">>, JObj),
+    maybe_save_account(Realm, TZ, AccountId, AuthToken),
+    AccountDb = wh_util:format_account_id(AccountId, 'encoded'),
+    Devices = get_devices_by_owner(AccountDb, OwnerId),
+    lists:foreach(
+      fun (DeviceId) ->
+              case cf_endpoint:get(DeviceId, AccountDb) of
+                  {'error', _E} ->
+                      lager:debug("no endpoint for device ~s: ~p", [DeviceId,_E]);
+                  {'ok', Endpoint} ->
+                      catch save_device(Realm, TZ, Endpoint, AuthToken)
+              end
+      end, Devices).
+
+-spec get_devices_by_owner(ne_binary(), ne_binary()) -> ne_binaries().
+get_devices_by_owner(AccountDb, OwnerId) ->
+    ViewOptions = [{'startkey', [OwnerId]}
+                  ,{'endkey', [OwnerId,wh_json:new()]}],
+    case couch_mgr:get_results(AccountDb, <<"cf_attributes/owned">>, ViewOptions) of
+        {'ok', JObjs} -> select_only_devices(JObjs);
+        {'error', _R} ->
+            lager:warning("unable to find documents owned by ~s: ~p", [OwnerId, _R]),
+            []
+    end.
+
+-spec select_only_devices([wh_json:object()]) -> ne_binaries().
+select_only_devices(JObjs) ->
+    lists:foldl(
+      fun (JObj, Acc) ->
+              case wh_json:get_value(<<"key">>, JObj) of
+                  [_OwnerId, <<"device">>] ->
+                      [wh_json:get_value(<<"value">>, JObj) | Acc];
+                  _ -> Acc
+              end
+      end, [], JObjs).
+
+-spec maybe_save_account(api_binary(), api_binary(), ne_binary(), ne_binary()) -> 'ok'.
+maybe_save_account('undefined', _, _, _) ->
+    laer:debug("cannot save account: realm is undefined");
+maybe_save_account(Realm, TZ, AccountId, AuthToken) ->
+    Setters = [ fun (J) -> wh_json:set_value(<<"realm">>, Realm, J) end
+              , fun (J) -> wh_json:set_value(<<"timezone">>, TZ, J) end ],
+    JObj = lists:foldl(fun (F,J) -> F(J) end, wh_json:new(), Setters),
```

Comment:
Why not wh_json:from_list?

JObj = wh_json:from_list([{<<"realm">>, Realm},
                                         ,{<<"timezone">>, Tz}]),